### PR TITLE
Naren/giacomo urban storm drains

### PR DESCRIPTION
**Pull request by nvohra0016**
New implementation which takes care of:
1. Different surface and pipe network domains. 
2. Pipe drain length evaluated as the difference of the bathymetry of surface and pipe network domain.

### CODE REVIEW

**Comment by gcapodag** [comment]
Thanks @nvohra0016 , will review ASAP.

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc [hunk 1]
```diff
@@ -69,32 +177,65 @@ void PipeDrainEvaluator::Evaluate_(const State& S,
...
   else {
      for (int c=0; c!=ncells; ++c) {
            res[0][c] = - mnhMask[0][c] *  2.0 / 3.0 * energ_loss_coeff_weir_ * mnhPerimeter * sqrtTwoG * pow(srfcDepth[0][c],3.0/2.0);
+           res[0][c] *= sink_source_coeff_;
      }
   }   
 
 }
 
+// Needs to be possibly UPDATED
```
Comment (gcapodag) [inline src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc:238]: Please update the `EvaluatePartialDerivative_` as well.

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh [hunk 2]
```diff
@@ -37,10 +38,14 @@ class PipeDrainEvaluator : public EvaluatorSecondaryMonotypeCV {
   virtual void EvaluatePartialDerivative_(const State& S,
           const Key& wrt_key, const Tag& wrt_tag,
           const std::vector<CompositeVector*>& result) override;
+  virtual void EnsureCompatibility_ToDeps_(State& S, const CompositeVectorSpace& fac) override;
```
Comment (gcapodag) [inline src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh:41]: Please change name to `EnsureDepsCompatibility_`

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh [hunk 3]
```diff
@@ -49,7 +54,11 @@ class PipeDrainEvaluator : public EvaluatorSecondaryMonotypeCV {
  double energ_loss_coeff_weir_; // weir 
  double energ_loss_coeff_subweir_; // submerged weir
  double energ_loss_coeff_orifice_; // orifice
- double drain_length_; // drain conduit length
+ double sink_source_coeff_; // coefficient that determines sink or source (when using same evaluator file for pipe or surface flow) 
+ bool pipe_flag, sw_flag, pipe_map_created, sw_map_created;
```
Comment (gcapodag) [inline src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh:58]: Please add `_` to the class variables at the end of their names, i.e. `pipe_flag_` for consistency

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc [hunk 4]
```diff
@@ -32,23 +32,126 @@ PipeDrainEvaluator::PipeDrainEvaluator(Teuchos::ParameterList& plist) :
...
+{
+  // grab the meshes 
+  Teuchos::RCP<const AmanziMesh::Mesh> pipe_mesh = S.GetMesh(pipe_domain_name_);
+  Teuchos::RCP<const AmanziMesh::Mesh> surface_mesh = S.GetMesh(sw_domain_name_);
+  
+  // get the manhole locations/ cell maps
+  Tag tag = my_keys_.front().second;
+
+  const Epetra_MultiVector& mnhMask = *S.GetPtr<CompositeVector>(mask_key_, tag)->ViewComponent("cell",false);
+  // loop over mesh cells and create map
```
Comment (gcapodag) [inline src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc:92]: Move comment one line down

**Review by gcapodag — CHANGES_REQUESTED** [review]
Thanks @nvohra0016 very good work. Just a few minor things before we merge.

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh (hunk 3 above)
Comment (nvohra0016) [reply src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh]: Done

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc (hunk 4 above)
Comment (nvohra0016) [reply src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc:92]: Done

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh (hunk 2 above)
Comment (nvohra0016) [reply src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh:41]: Hmm I don't think that would work? It is a virtual function.

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc (hunk 1 above)
Comment (nvohra0016) [reply src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.cc]: Done

File: src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh (hunk 2 above)
Comment (gcapodag) [reply src/pks/flow/constitutive_relations/sources/pipe_drain_evaluator.hh:41]: Sorry I thought it was a function you introduced

**Review by gcapodag — APPROVED** [review]

**Pull request merged**